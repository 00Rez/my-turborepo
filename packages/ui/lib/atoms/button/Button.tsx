import { useMemo } from 'react';
import { ButtonStyle, ButtonType } from 'devextreme/common';
import { default as DXButton } from 'devextreme-react/button';

import'./Button.scss'

export type ButtonProps = {
  disabled?: boolean;
  loading?: boolean;
  icon?: string;
  variant?: 'primary' | 'secondary' | 'danger' | 'success' | 'warning' | 'info';
  styling?: 'outlined' | 'filled' | 'text';
  onClick?: () => void;
} & React.PropsWithChildren;

export const Button = ({ 
  disabled,
  loading,
  icon,
  variant,
  styling,
  children,
  onClick
}: ButtonProps): React.ReactElement => {

  const buttonStyle: ButtonStyle = useMemo(() => {
    switch (styling) {
      case 'outlined':
        return 'outlined';
      case 'filled':
        return 'contained';
      case 'text':
        return 'text';
      default:
        return variant === 'secondary'
          || variant === 'info' 
          || variant === 'warning' ? 'text' : 'contained';
    }
  }, [styling, variant]);

  const type: ButtonType = useMemo(() => {
    switch (variant) {
      case 'primary':
        return 'default';
      case 'secondary':
        return 'default';
      case 'danger':
        return 'danger';
      case 'success':
        return 'success';
      case 'warning':
        return 'danger';
      case 'info':
        return 'default';
      default:
        return 'default';
    }
  }, [variant]);

  return (
    <DXButton
      disabled={disabled || loading}
      icon={icon}
      stylingMode={buttonStyle}
      type={type}
      onClick={onClick}
    >
      {children}
    </DXButton>
  )
}
