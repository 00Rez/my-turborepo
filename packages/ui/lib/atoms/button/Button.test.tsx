import { describe, it, expect, vi } from "vitest";
import { Button } from "./Button";
import { render } from "@testing-library/react";
import { PropsWithChildren } from "react";

vi.mock('devextreme-react/button', () => ({
  default: ({ children, ...props }: PropsWithChildren) => (
    <button {...props}>{children}</button>
  ),
}));

describe(Button.name, () => {
  it("renders", () => {
    const { container } = render(<Button>Example button</Button>);
    expect(container).toMatchSnapshot();
  });
});
