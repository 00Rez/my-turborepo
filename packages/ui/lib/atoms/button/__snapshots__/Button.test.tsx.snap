// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders 1`] = `
<div>
  <button
    class="_button_6ca4c3 _primary_6ca4c3 _filled_6ca4c3"
  >
    Example button
  </button>
</div>
`;
